pulpcore/cli/common/generic.py

```diff
@@ -42,15 +51,58 @@ def format_help_text(
             self.help = self.help.format(entity=entity_ctx.ENTITY, entities=entity_ctx.ENTITIES)
         super().format_help_text(ctx, formatter)
 
+    def get_params(self, ctx: click.Context) -> List[click.Parameter]:
+        params = super().get_params(ctx)
+        new_params: List[click.Parameter] = []
+        # Is it more efficient to go through and just remove options from params?
+        for param in params:
+            if isinstance(param, PulpOption):
+                if param.needs_contexts is not None:
+                    if not isinstance(ctx.obj, param.needs_contexts):
+                        continue
+            new_params.append(param)
+        return new_params
+
+
+class PulpGroup(click.Group):
+    def command(self, *args: Any, **kwargs: Any) -> Callable[[_FC], click.Command]:
+        kwargs["cls"] = kwargs.get("cls", PulpCommand)
+        return super().command(*args, **kwargs)
+
+    def group(self, *args: Any, **kwargs: Any) -> Callable[[_FC], click.Group]:
+        kwargs["cls"] = kwargs.get("cls", PulpGroup)
+        return super().group(*args, **kwargs)
+
+    def get_command(self, ctx: click.Context, cmd_name: str) -> Optional[click.Command]:
+        # Overwriting this removes the command, while overwriting list_commands just hides it
```

Comment:
Is `list_command` still listing it? Then we need to overwrite both.

Reply:
Oh no, my comment wasn't clear. `list_commands` calls `get_command` when listing out the commands in the help message. If you overwrite `list_commands` you can hide the command from the help message, but it will still be callable since `get_command` can still access it.

Reply:
Thanks for the clarification. That last step of calling `get_command` for everything in `list_command` was not obvious.